File: scala/plugins/Doppler.scala
// Doppler, as a voxgig/plugin definition.

package com.voxgig.sekreto.plugins

import scala.collection.immutable.ListMap

import com.voxgig.sekreto.*
import com.voxgig.sekreto.Providers.checkaddr

/** Doppler.
  *
  * The whole config is downloaded once - Doppler's own bulk endpoint - and
  * answered from memory, like a remote .env: `api.token` is the
  * `API_TOKEN` entry. A service token is config-scoped, so project and
  * config are only needed with broader tokens.
  */
class Doppler(
    token: Option[String] = None,
    project: Option[String] = None,
    config: Option[String] = None,
    addr: Option[String] = None,
) extends Provider:

  private var values: Option[Map[String, String]] = None

  private def load(): Map[String, String] =
    values match
      case Some(loaded) => loaded
      case None =>
        val useaddr = trimslash(first(addr, Some("https://api.doppler.com")))
        checkaddr(useaddr)

        var url = s"$useaddr/v3/configs/config/secrets/download?format=json"
        project.filter(_.nonEmpty).foreach(value => url += "&project=" + uriescape(value))
        config.filter(_.nonEmpty).foreach(value => url += "&config=" + uriescape(value))

        val res = fetchjson(
          "GET",
          url,
          Map("authorization" -> s"Bearer ${token.getOrElse("")}"),
        )

        val body = res.body.asobj
        if 200 != res.status || body.isEmpty then
          throw SekretoError(s"sekreto: doppler error: ${res.status}")

        var loaded = ListMap.empty[String, String]
        for (key, value) <- body.get do value.text.foreach(text => loaded = loaded.updated(key, text))

        values = Some(loaded)
        loaded

  override def lookup(name: String): Option[String] = load().get(envkey(name))

  override def describe(): String =
    "doppler" + (
      if project.exists(_.nonEmpty) then s":${project.get}/${config.getOrElse("")}" else ""
    )

/** The `doppler` provider kind, as a voxgig/plugin definition. A consumer
  * imports this and hands it to `Sekreto`; a consumer that does not gets a
  * `Sekreto` that cannot build a chain entry of that kind.
  */
val doppler: Definition = providerplugin("doppler", spec =>
  Doppler(spec.token, spec.project, spec.config, spec.addr))

File: kotlin/plugins/Doppler.kt
// Doppler, as a voxgig/plugin definition.

package com.voxgig.sekreto.plugins

import com.voxgig.sekreto.Definition
import com.voxgig.sekreto.Provider
import com.voxgig.sekreto.Providers.checkaddr
import com.voxgig.sekreto.SekretoError
import com.voxgig.sekreto.envkey
import com.voxgig.sekreto.providerplugin

/**
 * Doppler.
 *
 * The whole config is downloaded once - Doppler's own bulk endpoint - and
 * answered from memory, like a remote .env: `api.token` is the
 * `API_TOKEN` entry. A service token is config-scoped, so project and
 * config are only needed with broader tokens.
 */
class Doppler(
    private val token: String? = null,
    private val project: String? = null,
    private val config: String? = null,
    private val addr: String? = null,
) : Provider {

    private var values: Map<String, String>? = null

    private fun load(): Map<String, String> {
        values?.let { return it }

        val useaddr = trimslash(first(addr, "https://api.doppler.com"))
        checkaddr(useaddr)

        var url = "$useaddr/v3/configs/config/secrets/download?format=json"
        if (!project.isNullOrEmpty()) {
            url += "&project=" + uriescape(project)
        }
        if (!config.isNullOrEmpty()) {
            url += "&config=" + uriescape(config)
        }

        val res = fetchjson(
            "GET",
            url,
            mapOf("authorization" to "Bearer ${token ?: ""}"),
        )

        val body = res.body?.asobj
        if (200 != res.status || null == body) {
            throw SekretoError("sekreto: doppler error: ${res.status}")
        }

        val loaded = LinkedHashMap<String, String>()
        for ((key, value) in body) {
            value.text?.let { loaded[key] = it }
        }

        values = loaded
        return loaded
    }

    override fun lookup(name: String): String? = load()[envkey(name)]

    override fun describe(): String =
        "doppler" + if (project.isNullOrEmpty()) "" else ":$project/${config ?: ""}"
}

/** The `doppler` provider kind, as a voxgig/plugin definition. A
 * consumer imports this and hands it to `Sekreto`; a consumer that does
 * not is a `Sekreto` that cannot build a `doppler` chain entry. */
val doppler: Definition = providerplugin("doppler") { spec ->
    Doppler(spec.token, spec.project, spec.config, spec.addr)
}
